// @nuxt/vue-app-edge/dist/vue-app.js
/*!
 * @nuxt/vue-app-edge v2.16.0-27720022.54e852f (c) 2016-2022
 * Released under the MIT License
 * Repository: https://github.com/nuxt/nuxt.js
 * Website: https://nuxtjs.org
*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const path = require('path');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

const path__default = /*#__PURE__*/_interopDefaultLegacy(path);

var dependencies = {
	"node-fetch-native": "^0.1.4",
	ufo: "^0.8.5",
	unfetch: "^4.2.0",
	vue: "^2.7.10",
	"vue-client-only": "^2.1.0",
	"vue-meta": "^2.4.0",
	"vue-no-ssr": "^1.1.1",
	"vue-router": "^3.6.4",
	"vue-template-compiler": "^2.7.10",
	vuex: "^3.6.2"
};

const template = {
  dependencies,
  dir: path__default["default"].join(__dirname, "..", "template"),
  files: [
    "App.js",
    "client.js",
    "index.js",
    "jsonp.js",
    "router.js",
    "router.scrollBehavior.js",
    "routes.json",
    "server.js",
    "utils.js",
    "empty.js",
    "mixins/fetch.server.js",
    "mixins/fetch.client.js",
    "components/nuxt-error.vue",
    "components/nuxt-child.js",
    "components/nuxt-link.server.js",
    "components/nuxt-link.client.js",
    "components/nuxt.js",
    "views/app.template.html",
    "views/error.html"
  ]
};

exports.template = template;
